--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -73,7 +73,6 @@
 EXPORT_SYMBOL(memory_cgrp_subsys);
 
 struct mem_cgroup *root_mem_cgroup __read_mostly;
-EXPORT_SYMBOL_GPL(root_mem_cgroup);
 
 /* Active memory cgroup to use from an interrupt context */
 DEFINE_PER_CPU(struct mem_cgroup *, int_active_memcg);
@@ -859,7 +858,6 @@
 	if (!mem_cgroup_disabled())
 		__mod_memcg_lruvec_state(lruvec, idx, val);
 }
-EXPORT_SYMBOL_GPL(__mod_lruvec_state);
 
 void __mod_lruvec_slab_state(void *p, enum node_stat_item idx, int val)
 {
@@ -1377,38 +1375,6 @@
 	return lruvec;
 }
 
-struct lruvec *page_to_lruvec(struct page *page, pg_data_t *pgdat)
-{
-	struct lruvec *lruvec;
-
-	lruvec = mem_cgroup_page_lruvec(page, pgdat);
-
-	return lruvec;
-}
-EXPORT_SYMBOL_GPL(page_to_lruvec);
-
-void do_traversal_all_lruvec(void)
-{
-	pg_data_t *pgdat;
-
-	for_each_online_pgdat(pgdat) {
-		struct mem_cgroup *memcg = NULL;
-
-		spin_lock_irq(&pgdat->lru_lock);
-		memcg = mem_cgroup_iter(NULL, NULL, NULL);
-		do {
-			struct lruvec *lruvec = mem_cgroup_lruvec(memcg, pgdat);
-
-			trace_android_vh_do_traversal_lruvec(lruvec);
-
-			memcg = mem_cgroup_iter(NULL, memcg, NULL);
-		} while (memcg);
-
-		spin_unlock_irq(&pgdat->lru_lock);
-	}
-}
-EXPORT_SYMBOL_GPL(do_traversal_all_lruvec);
-
 /**
  * mem_cgroup_update_lru_size - account for adding or removing an lru page
  * @lruvec: mem_cgroup per zone lru vector
@@ -1447,7 +1413,6 @@
 	if (nr_pages > 0)
 		*lru_size += nr_pages;
 }
-EXPORT_SYMBOL_GPL(mem_cgroup_update_lru_size);
 
 /**
  * mem_cgroup_margin - calculate chargeable space of a memory cgroup
@@ -2921,6 +2886,7 @@
 	 * - LRU isolation
 	 * - lock_page_memcg()
 	 * - exclusive reference
+	 * - mem_cgroup_trylock_pages()
 	 */
 	page->mem_cgroup = memcg;
 }
@@ -5328,6 +5294,7 @@
 
 static void mem_cgroup_free(struct mem_cgroup *memcg)
 {
+	lru_gen_exit_memcg(memcg);
 	memcg_wb_domain_exit(memcg);
 	/*
 	 * Flush percpu vmstats and vmevents to guarantee the value correctness
@@ -5402,6 +5369,7 @@
 	spin_lock(&memcg_idr_lock);
 	idr_replace(&mem_cgroup_idr, memcg, memcg->id.id);
 	spin_unlock(&memcg_idr_lock);
+	lru_gen_init_memcg(memcg);
 	trace_android_vh_mem_cgroup_alloc(memcg);
 	return memcg;
 fail:
@@ -6297,6 +6265,30 @@
 }
 #endif
 
+#ifdef CONFIG_LRU_GEN
+static void mem_cgroup_attach(struct cgroup_taskset *tset)
+{
+	struct task_struct *task;
+	struct cgroup_subsys_state *css;
+
+	/* find the first leader if there is any */
+	cgroup_taskset_for_each_leader(task, css, tset)
+		break;
+
+	if (!task)
+		return;
+
+	task_lock(task);
+	if (task->mm && READ_ONCE(task->mm->owner) == task)
+		lru_gen_migrate_mm(task->mm);
+	task_unlock(task);
+}
+#else
+static void mem_cgroup_attach(struct cgroup_taskset *tset)
+{
+}
+#endif /* CONFIG_LRU_GEN */
+
 /*
  * Cgroup retains root cgroups across [un]mount cycles making it necessary
  * to verify whether we're attached to the default hierarchy on each mount
@@ -6649,6 +6641,7 @@
 	.css_free = mem_cgroup_css_free,
 	.css_reset = mem_cgroup_css_reset,
 	.can_attach = mem_cgroup_can_attach,
+	.attach = mem_cgroup_attach,
 	.cancel_attach = mem_cgroup_cancel_attach,
 	.post_attach = mem_cgroup_move_task,
 	.bind = mem_cgroup_bind,
