--- a/fs/kernfs/file.c
+++ b/fs/kernfs/file.c
@@ -39,6 +39,15 @@
 	struct list_head	files; /* goes through kernfs_open_file.list */
 };
 
+static struct kmem_cache *kmem_open_node_pool;
+static struct kmem_cache *kmem_open_file_pool;
+
+void __init init_kernfs_file_pool(void)
+{
+	kmem_open_node_pool = KMEM_CACHE(kernfs_open_node, SLAB_HWCACHE_ALIGN | SLAB_PANIC);
+	kmem_open_file_pool = KMEM_CACHE(kernfs_open_file, SLAB_HWCACHE_ALIGN | SLAB_PANIC);
+}
+
 /*
  * kernfs_notify() may be called from any context and bounces notifications
  * through a work item.  To minimize space overhead in kernfs_node, the
@@ -257,6 +266,7 @@
 	struct kernfs_open_file *of = kernfs_of(iocb->ki_filp);
 	ssize_t len = iov_iter_count(iter);
 	const struct kernfs_ops *ops;
+	char buf_onstack[SZ_64] __aligned(sizeof(long));
 	char *buf;
 
 	if (of->atomic_write_len) {
@@ -266,13 +276,17 @@
 		len = min_t(size_t, len, PAGE_SIZE);
 	}
 
-	buf = of->prealloc_buf;
-	if (buf)
-		mutex_lock(&of->prealloc_mutex);
-	else
-		buf = kmalloc(len + 1, GFP_KERNEL);
-	if (!buf)
-		return -ENOMEM;
+	if (len < sizeof(buf_onstack)) {
+		buf = buf_onstack;
+	} else {
+		buf = of->prealloc_buf;
+		if (buf)
+			mutex_lock(&of->prealloc_mutex);
+		else
+			buf = kmalloc(len + 1, GFP_KERNEL);
+		if (!buf)
+			return -ENOMEM;
+	}
 
 	if (copy_from_iter(buf, len, iter) != len) {
 		len = -EFAULT;
@@ -306,7 +320,7 @@
 out_free:
 	if (buf == of->prealloc_buf)
 		mutex_unlock(&of->prealloc_mutex);
-	else
+	else if (buf != buf_onstack)
 		kfree(buf);
 	return len;
 }
@@ -544,12 +558,13 @@
 	mutex_unlock(&kernfs_open_file_mutex);
 
 	if (on) {
-		kfree(new_on);
+		if (new_on)
+			kmem_cache_free(kmem_open_node_pool, new_on);
 		return 0;
 	}
 
 	/* not there, initialize a new one and retry */
-	new_on = kmalloc(sizeof(*new_on), GFP_KERNEL);
+	new_on = kmem_cache_alloc(kmem_open_node_pool, GFP_KERNEL);
 	if (!new_on)
 		return -ENOMEM;
 
@@ -591,7 +606,8 @@
 	spin_unlock_irqrestore(&kernfs_open_node_lock, flags);
 	mutex_unlock(&kernfs_open_file_mutex);
 
-	kfree(on);
+	if (on)
+		kmem_cache_free(kmem_open_node_pool, on);
 }
 
 static int kernfs_fop_open(struct inode *inode, struct file *file)
@@ -625,7 +641,7 @@
 
 	/* allocate a kernfs_open_file for the file */
 	error = -ENOMEM;
-	of = kzalloc(sizeof(struct kernfs_open_file), GFP_KERNEL);
+	of = kmem_cache_zalloc(kmem_open_file_pool, GFP_KERNEL);
 	if (!of)
 		goto err_out;
 
@@ -716,7 +732,7 @@
 	seq_release(inode, file);
 err_free:
 	kfree(of->prealloc_buf);
-	kfree(of);
+	kmem_cache_free(kmem_open_file_pool, of);
 err_out:
 	kernfs_put_active(kn);
 	return error;
@@ -760,7 +776,8 @@
 	kernfs_put_open_node(kn, of);
 	seq_release(inode, filp);
 	kfree(of->prealloc_buf);
-	kfree(of);
+	if (of)
+		kmem_cache_free(kmem_open_file_pool, of);
 
 	return 0;
 }
